## Lab Assignment 10: <br>
### Student Name:Fatimah Almusawi
### Student ID:900400

This assignment targets to assure that the students understood basic IR concepts.<br>

**Part 1: Implementing an IR System (10/16)**<br>

Consider a collection of 1000 documents and a set of 10 queries. Implement an IR system based on the Vector Space Model (VSM) using TF-IDF weighting.<br>

Dataset:<br>

- 1000 documents (text content for each document)
- 10 sample queries <br>
IR System:<br>

Implement TF-IDF calculation for document-term matrix construction. <br>
Develop a cosine similarity-based retrieval system to rank documents for each query. <br>
Rank the top 10 documents for each query using the IR system. <br>
Evaluation: <br>

Compute Precision at k (P@k) for k=5, k=6, and k=10 for each query.<br>
Calculate Mean Average Precision (MAP) across all queries. <br>
Calculate the Mean Reciprocal Rank (MRR) across all queries.(3) <br><br>
**Part 2: Assessing Inter-Annotator Agreement (6/16)**<br>

Given the relevance assessments by three different annotators for a set of documents:<br>

Annotator 1,2 and 3 relevance assessments<br>

You are expected to: <br>

Compute pairwise Cohen's Kappa values for the annotators' relevance assessments.<br>
Discuss the observed agreement levels among annotators.<br>
Explain how to improve the kappa value if we are not satisfied with the kappa.<br>
Hint: Check Kappa Measure Last Slide Week 11


----------------------------------------------------------------------------------------
- Part 1

In [12]:
documents = [
    "Machine learning is transforming various industries.",
    "Natural language processing helps in text analysis.",
    "AI algorithms can improve decision-making processes.",
    "Data science involves extracting insights from data.",
    "Robotics is a field combining hardware and software.",
    "Deep learning models require large amounts of data.",
    "Blockchain technology secures digital transactions.",
    "Cloud computing offers scalable computing power.",
    "Virtual reality provides immersive experiences.",
    "Augmented reality enhances real-world environments.",
    "Internet of Things connects devices for data exchange.",
    "Biometric authentication ensures secure access.",
    "Quantum computing promises faster computations.",
    "Cybersecurity protects against digital threats.",
    "Ethical considerations in AI development are crucial.",
    "Healthcare benefits from AI-driven diagnostics.",
    "E-commerce relies on personalized recommendation systems.",
    "Autonomous vehicles revolutionize transportation.",
    "Smart cities utilize IoT for efficient operations.",
    "Social media analysis aids in understanding trends.",
    "Predictive analytics anticipates future outcomes.",
    "Remote work trends increase reliance on digital tools.",
    "Energy efficiency through smart grids is vital.",
    "Fintech innovations reshape financial services.",
    "Data privacy concerns arise in the era of big data.",
    "Artificial general intelligence remains a challenge.",
    "Human-computer interaction shapes user experiences.",
    "Big data analytics drives informed decision-making.",
    "Climate change predictions benefit from AI models.",
    "Bioinformatics applies computational techniques to biology.",
    "Robotic process automation streamlines workflows.",
    "Industry 4.0 integrates AI with manufacturing.",
    "Sentiment analysis detects emotions in text data.",
    "Reinforcement learning powers autonomous agents.",
    "Data visualization simplifies complex information.",
    "Edge computing reduces latency in data processing.",
    "Smart farming optimizes agricultural practices.",
    "AI in education enhances personalized learning.",
    "Natural disaster predictions leverage machine learning.",
    "Media recommendation systems personalize content.",
    "Speech recognition enables hands-free interactions.",
    "Genetic algorithms mimic natural selection.",
    "Neural networks simulate the human brain.",
    "Behavioral analytics uncovers patterns in behavior.",
    "Spatial analysis benefits from geospatial data.",
    "Explainable AI improves transparency in models.",
    "AI ethics guide responsible technology development.",
    "Prescriptive analytics offers actionable insights.",
    "Supply chain optimization employs predictive models.",
    "Emotion AI detects emotions in facial expressions.",
    "Distributed ledger technology ensures secure transactions.",
    "Biological data analysis aids in medical research.",
    "Smart grid technology enhances energy distribution.",
    "AI-powered chatbots automate customer support.",
    "Machine translation breaks language barriers.",
    "Personalized medicine tailors treatments to individuals.",
    "Data-driven marketing optimizes customer engagement.",
    "Smart home devices enhance living experiences.",
    "Time series forecasting predicts future trends.",
    "AI-driven creativity challenges human capabilities.",
    "Intelligent document processing automates data extraction.",
    "Graph analytics explores relationships in networks.",
    "Adversarial attacks pose challenges to AI security.",
    "Predictive maintenance minimizes equipment downtime.",
    "Spatial reasoning enhances AI navigation systems.",
    "Privacy-preserving techniques protect sensitive data.",
    "Automated decision-making raises ethical concerns.",
    "Behavioral biometrics verifies user identities.",
    "Explainable recommendations increase user trust.",
    "Quantum machine learning explores quantum states.",
    "Internet censorship detection uses AI techniques.",
    "Smart wearables monitor health and fitness.",
    "Intelligent tutoring systems adapt to student needs.",
    "Graph neural networks model complex relationships.",
    "AI in sports analytics improves performance analysis.",
    "Facial recognition technology raises privacy debates.",
    "Digital twin technology replicates physical systems.",
    "Emotion detection in video content aids analysis.",
    "Neuromorphic computing mimics brain structure.",
    "Robotic surgery enhances precision in operations.",
    "AI-powered language translation assists global communication.",
    "Automated content moderation filters online content.",
    "Network anomaly detection identifies security threats.",
    "Predictive policing uses data to prevent crime.",
    "Quantum cryptography ensures secure communications.",
    "AI-generated art challenges traditional creativity.",
    "Ethical considerations in autonomous vehicles are debated.",
    "Adaptive learning systems personalize educational content.",
    "Biomedical image analysis aids in disease diagnosis.",
    "Explainable computer vision interprets image features.",
    "AI-driven personalization enhances user experiences.",
    "Conversational AI improves human-like interactions.",
    "Federated learning protects individual data privacy.",
    "Human-centered AI design prioritizes user needs.",
    "Quantum annealing solves optimization problems.",
    "Behavior-based authentication enhances security.",
    "AI in music composition transforms creative processes.",
    "Semantic search improves information retrieval.",
    "Recommender systems optimize user preferences.",
    "Information Retrieval systems are outdated without neural networks"
]
print(len(documents))

100


In [33]:
queries = [
    "What are the impacts of AI on healthcare?",
    "How does machine learning improve financial services?",
    "What is the role of AI in autonomous vehicles?",
    "Explain the applications of natural language processing.",
    "How does robotics benefit from AI integration?",
    "What are the challenges in implementing AI in education?",
    "How does data science contribute to climate change predictions?",
    "What are the ethical considerations in AI development?",
    "Explain the significance of AI in smart cities.",
    "How does AI enhance cybersecurity measures?"
]
print(len(queries))

10


In [97]:
ground_truth =[[3, 74, 38, 92, 71, 72, 98, 48, 18, 100], 
                [12, 26, 52, 54, 21, 63, 64, 45, 14, 83],
                [58, 77, 62, 31, 6, 61, 36, 96, 85, 18], 
                [99, 62, 98, 34, 63, 79, 43, 31, 3, 16],
                [8, 88, 73, 82, 37, 25, 34, 87, 66, 58], 
                [90, 40, 84, 64, 34, 2, 73, 23, 59, 89],
                [80, 10, 86, 21, 68, 37, 83, 57, 6, 98],
                [22, 3, 2, 44, 56, 80, 50, 63, 87, 13],
                [1, 18, 22, 44, 99, 72, 24, 95, 10, 87],
                [31, 10, 56, 50, 75, 4, 18, 85, 84, 74]
                           ]
print(len(ground_truth))


10


In [102]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.metrics import cohen_kappa_score
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
import numpy as np


Step 1: Data Loading

First, we load our dataset comprising 1000 documents. This dataset will serve as the basis for our Information Retrieval (IR) system. The documents are loaded into a list where each entry represents a document.

In [5]:
## TODO: Implement your solution for part 1
# Assuming documents are stored in a list called 'documents'
documents = ["Document 1 text", "Document 2 text", ..., "Document 1000 text"]


### step 2: Tokenization and removing stopwords
Here, we tokenize our documents, breaking them down into individual words. This step is crucial for analyzing the text at the word level.


In [15]:
from nltk.tokenize import word_tokenize
from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.corpus import stopwords


# Tokenize documents
tokenized_documents = [word_tokenize(document) for document in documents]

# Define the list of English stopwords
english_stopwords = stopwords.words('english')

# Preprocess the tokenized documents: remove stopwords and convert to lowercase
cleaned_data = [[word.lower() for word in document if word.lower() not in english_stopwords] for document in tokenized_documents]

# Join tokens back into sentences
cleaned_sentences = [' '.join(document) for document in cleaned_data]
cleaned_sentences



['machine learning transforming various industries .',
 'natural language processing helps text analysis .',
 'ai algorithms improve decision-making processes .',
 'data science involves extracting insights data .',
 'robotics field combining hardware software .',
 'deep learning models require large amounts data .',
 'blockchain technology secures digital transactions .',
 'cloud computing offers scalable computing power .',
 'virtual reality provides immersive experiences .',
 'augmented reality enhances real-world environments .',
 'internet things connects devices data exchange .',
 'biometric authentication ensures secure access .',
 'quantum computing promises faster computations .',
 'cybersecurity protects digital threats .',
 'ethical considerations ai development crucial .',
 'healthcare benefits ai-driven diagnostics .',
 'e-commerce relies personalized recommendation systems .',
 'autonomous vehicles revolutionize transportation .',
 'smart cities utilize iot efficient oper

### Step 3: TF-IDF Calculation
We convert our clean text data into TF-IDF vectors. This numerical transformation emphasizes the importance of words across documents and queries.


In [14]:
# Define the TF-IDF Vectorizer and transform the cleaned sentences
tfidf_vectorizer = TfidfVectorizer()
tfidf_matrix = tfidf_vectorizer.fit_transform(cleaned_sentences)

# Show the TF-IDF matrix
print(tfidf_matrix)

  (0, 163)	0.49120372384880856
  (0, 327)	0.49120372384880856
  (0, 312)	0.49120372384880856
  (0, 181)	0.34109961564339564
  (0, 186)	0.3997596243233148
  (1, 11)	0.3259493911700577
  (1, 305)	0.43071911294080617
  (1, 151)	0.4693864999729862
  (1, 242)	0.4032841652344177
  (1, 178)	0.4032841652344177
  (1, 202)	0.4032841652344177
  (2, 241)	0.43630037870457955
  (2, 188)	0.4085099284682032
  (2, 81)	0.4085099284682032
  (2, 158)	0.47546881841107386
  (2, 9)	0.43630037870457955
  (2, 7)	0.24382911739385216
  (3, 168)	0.40332998980704693
  (3, 123)	0.4395385451021021
  (3, 176)	0.4395385451021021
  (3, 273)	0.4395385451021021
  (3, 78)	0.5076834854816823
  (4, 289)	0.4472135954999579
  (4, 148)	0.4472135954999579
  (4, 56)	0.4472135954999579
  :	:
  (95, 23)	0.4623932891322133
  (95, 111)	0.3619774719865355
  (96, 71)	0.4425900318406722
  (96, 313)	0.4425900318406722
  (96, 61)	0.4425900318406722
  (96, 201)	0.4425900318406722
  (96, 241)	0.40613009944220513
  (96, 7)	0.226968273527723

In [77]:


# Tokenize, clean, and prepare queries similarly to documents
tokenized_queries = [word_tokenize(query) for query in queries]
cleaned_queries = [
    ' '.join([word.lower() for word in query if word.lower() not in english_stopwords])
    for query in tokenized_queries
]

# Transform the cleaned queries to TF-IDF vectors using the same vectorizer
tfidf_queries = tfidf_vectorizer.transform(cleaned_queries)
print(tfidf_queries)


  (0, 150)	0.8898182575679641
  (0, 7)	0.45631509782025886
  (1, 283)	0.49120372384880856
  (1, 186)	0.3997596243233148
  (1, 181)	0.34109961564339564
  (1, 158)	0.49120372384880856
  (1, 132)	0.49120372384880856
  (2, 328)	0.675893887742558
  (2, 28)	0.6328423655134223
  (2, 7)	0.37772740557659634
  (3, 242)	0.5773502691896257
  (3, 202)	0.5773502691896257
  (3, 178)	0.5773502691896257
  (4, 271)	0.6647515987814472
  (4, 33)	0.6647515987814472
  (4, 7)	0.34089679352410995
  (5, 100)	0.7069016836152037
  (5, 50)	0.6073507768032713
  (5, 7)	0.3625121289259898
  (6, 273)	0.4893761419033552
  (6, 230)	0.44906203682282747
  (6, 78)	0.2826238883956818
  (6, 54)	0.4893761419033552
  (6, 51)	0.4893761419033552
  (7, 116)	0.524313122582309
  (7, 86)	0.559981527989344
  (7, 68)	0.559981527989344
  (7, 7)	0.3129490790998236
  (8, 287)	0.552778160814236
  (8, 53)	0.7415108690124079
  (8, 7)	0.3802603529994866
  (9, 110)	0.6647515987814472
  (9, 77)	0.6647515987814472
  (9, 7)	0.34089679352410995


### Step 4: Cosine Similarity Calculation
With our documents as TF-IDF vectors, we now calculate the cosine similarity between each document and each query. This similarity score helps us gauge the relevance of documents to queries.
before calculating the cosine similarity,I will preprocessing the queries:

In [50]:

# Function to preprocess and transform a single query
def preprocess_and_transform_query(query):
    query_tokens = word_tokenize(query)
    query_cleaned = [word.lower() for word in query_tokens if word.lower() not in english_stopwords]
    query_cleaned_combined = ' '.join(query_cleaned)
    query_tfIdf_vector = tfidf_vectorizer.transform([query_cleaned_combined])
    return query_tfIdf_vector

# Calculate cosine similarities for each query
for query in queries:
    query_tfIdf_vector = preprocess_and_transform_query(query)
    cosine_similarities = cosine_similarity(query_tfIdf_vector, tfidf_matrix)
    print(f"Query: {query}")
    print("Cosine Similarities:", cosine_similarities)


Query: What are the impacts of AI on healthcare?
Cosine Similarities: [[0.         0.         0.11126291 0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.12189804 0.57900347 0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.10717419 0.
  0.         0.1295451  0.         0.         0.         0.
  0.         0.13511749 0.         0.         0.         0.
  0.         0.         0.         0.12831544 0.10839286 0.
  0.         0.10873925 0.         0.         0.         0.10520966
  0.         0.         0.         0.         0.         0.11544205
  0.         0.         0.10749035 0.         0.11552197 0.
  0.         0.         0.         0.         0.11126291 0.
  0.         0.         0.11754444 0.         0.         0.
  0.         0.         0.09814197 0.         0.         0.
  0.         0.10633167 0.         0.         0.         0.
  0.12232735 0

### Ranking the document according to the cosine similarity
Ranking top 10 for each query:

In [90]:
# Assuming you have a list 'queries' that contains the text of each query
# Generate query names like "Query 1", "Query 2", ...
query_names = [f"Query {i+1}" for i in range(len(queries))]

# Now, modify the print statements to include query names
for query_index, top_10_indices in enumerate(top_10_ranked_docs_per_query):
    # Use the query name from the query_names list
    query_name = query_names[query_index]
    print(f"{query_name}: {queries[query_index]}")
    print("Top 10 documents ranked by cosine similarity:")
    for rank, doc_index in enumerate(top_10_indices, start=1):
        print(f"\t{rank}. Document Index: {doc_index} - Cosine Similarity: {cosine_similarities[query_index, doc_index]:.3f}")
    print("\n" + "-"*50 + "\n")


Query 1: What are the impacts of AI on healthcare?
Top 10 documents ranked by cosine similarity:
	1. Document Index: 15 - Cosine Similarity: 0.579
	2. Document Index: 37 - Cosine Similarity: 0.135
	3. Document Index: 31 - Cosine Similarity: 0.130
	4. Document Index: 45 - Cosine Similarity: 0.128
	5. Document Index: 90 - Cosine Similarity: 0.122
	6. Document Index: 14 - Cosine Similarity: 0.122
	7. Document Index: 74 - Cosine Similarity: 0.118
	8. Document Index: 64 - Cosine Similarity: 0.116
	9. Document Index: 59 - Cosine Similarity: 0.115
	10. Document Index: 91 - Cosine Similarity: 0.112

--------------------------------------------------

Query 2: How does machine learning improve financial services?
Top 10 documents ranked by cosine similarity:
	1. Document Index: 23 - Cosine Similarity: 0.439
	2. Document Index: 0 - Cosine Similarity: 0.276
	3. Document Index: 38 - Cosine Similarity: 0.259
	4. Document Index: 69 - Cosine Similarity: 0.237
	5. Document Index: 2 - Cosine Similarity

### Evaluation:<br>
#### Precision at k (P@k) for k=5, k=6, and k=10 for each query.
#### Mean Average Precision (MAP) across all queries. 
#### the Mean Reciprocal Rank (MRR) across all queries.

In [91]:
def precision_at_k(relevant_indices, retrieved_indices, k):
    relevant_set = set(relevant_indices)
    retrieved_set = set(retrieved_indices[:k])
    relevant_and_retrieved = len(relevant_set.intersection(retrieved_set))
    return relevant_and_retrieved / k


In [92]:
def average_precision(relevant_indices, retrieved_indices):
    ap_sum = 0
    hit_count = 0
    for i, doc_index in enumerate(retrieved_indices):
        if doc_index in relevant_indices:
            hit_count += 1
            precision = hit_count / (i + 1)
            ap_sum += precision
    return ap_sum / len(relevant_indices) if relevant_indices else 0

def mean_average_precision(relevant_indices_list, retrieved_indices_list):
    aps = [average_precision(relevant, retrieved) for relevant, retrieved in zip(relevant_indices_list, retrieved_indices_list)]
    return sum(aps) / len(aps)


In [93]:
def mean_reciprocal_rank(relevant_indices_list, retrieved_indices_list):
    reciprocal_ranks = []
    for relevant_indices, retrieved_indices in zip(relevant_indices_list, retrieved_indices_list):
        for rank, doc_index in enumerate(retrieved_indices, start=1):
            if doc_index in relevant_indices:
                reciprocal_ranks.append(1 / rank)
                break
    return sum(reciprocal_ranks) / len(reciprocal_ranks) if reciprocal_ranks else 0


In [101]:

#here i transferred the details of the top 10 ranked documents per query to a list
top_10_ranked_docs_per_query = [
    [15, 37, 31, 45, 90, 14, 74, 64, 59, 91],

     [23, 0, 38, 69, 2, 54, 37, 33, 92, 87],

    [17, 86, 33, 37, 31, 45, 90, 14, 15, 74],

    [1, 54, 41, 38, 35, 60, 80, 27, 28, 29],
   
    [28, 4, 37, 31, 45, 90, 14, 15, 74, 64],
    
    [37, 59, 62, 85, 31, 45, 90, 14, 15, 74],
    
    [28, 3, 38, 24, 44, 34, 92, 32, 51, 56],
   
    [14, 86, 46, 66, 37, 31, 45, 90, 15, 74],
    
    [18, 36, 22, 52, 71, 57, 37, 31, 45, 90],
    
    [13, 57, 37, 31, 45, 90, 14, 15, 74, 64]
]

# Calculate P@k for k=5, k=6, and k=10
for k in [5, 6, 10]:
    p_at_k_values = [precision_at_k(ground_truth[i], top_10_ranked_docs_per_query[i], k) for i in range(len(ground_truth))]
    print(f"Average P@{k}: {sum(p_at_k_values) / len(p_at_k_values):.3f}")

# Calculate MAP
map_score = mean_average_precision(ground_truth, top_10_ranked_docs_per_query)
print(f"MAP: {map_score:.3f}")

# Calculate MRR
mrr_score = mean_reciprocal_rank(ground_truth, top_10_ranked_docs_per_query)
print(f"MRR: {mrr_score:.3f}")


Average P@5: 0.120
Average P@6: 0.117
Average P@10: 0.100
MAP: 0.038
MRR: 0.370


**Precision at k (P@k)**
P@5: 0.120 - On average, 12% of the top 5 documents retrieved for each query are relevant.

P@6: 0.117 - Extending to the top 6 documents, the precision slightly decreases to 11.7%, indicating that the additional documents at the 6th position are often not relevant.

P@10: 0.100 - When considering the top 10 documents, the average precision further decreases to 10%. This suggests that as more documents are considered, the proportion of relevant documents within the top ranks decreases.

**Mean Average Precision (MAP): 0.038**
The MAP score of 0.038 is quite low, indicating that on average, the system does not rank relevant documents highly across all queries.

**Mean Reciprocal Rank (MRR): 0.370**
An MRR of 0.370 indicates that on average, the first relevant document appears relatively high in the ranked list across queries. The MRR score is This score is about how quickly we find the first good document, not how many good documents there are in total.







-------------------------------------------------------------------
- Part 2

In [74]:
# 0 is not relevant, 1 is relevant
annotator1_relevance = [1, 0, 1, 0, 1, 1, 0, 0, 1, 0]  
annotator2_relevance = [1, 1, 1, 0, 1, 0, 0, 1, 1, 1]  
annotator3_relevance = [1, 0, 0, 0, 1, 0, 0, 1, 1, 0]  

In [75]:
## TODO: Implement your solution for part 2



# Compute pairwise Cohen's Kappa scores
kappa_1_2 = cohen_kappa_score(annotator1_relevance, annotator2_relevance)
kappa_1_3 = cohen_kappa_score(annotator1_relevance, annotator3_relevance)
kappa_2_3 = cohen_kappa_score(annotator2_relevance, annotator3_relevance)

print(f"Kappa (Annotator 1 vs. Annotator 2): {kappa_1_2:.3f}")
print(f"Kappa (Annotator 1 vs. Annotator 3): {kappa_1_3:.3f}")
print(f"Kappa (Annotator 2 vs. Annotator 3): {kappa_2_3:.3f}")


Kappa (Annotator 1 vs. Annotator 2): 0.200
Kappa (Annotator 1 vs. Annotator 3): 0.400
Kappa (Annotator 2 vs. Annotator 3): 0.444


**Discussing the Observed Agreement Levels**<br>
● Kappa >0.8  good agreement
●  0.67<Kappa<0.8  Tentative Conclusions
●  >2 judges: average pairwise 𝑘𝑎𝑝𝑝𝑎


- Kappa (Annotator 1 vs. Annotator 2) = 0.200: This indicates slight agreement beyond chance between Annotator 1 and Annotator 2. The low score suggests that the criteria for judging relevance might not be consistently understood or applied between these two annotators.

- Kappa (Annotator 1 vs. Annotator 3) = 0.400: This score falls into the "fair agreement" category, suggesting that while Annotator 1 and Annotator 3 have some level of consistent application of the relevance criteria, there's still a significant room for improvement.

- Kappa (Annotator 2 vs. Annotator 3) = 0.444: This is close to moderate agreement but still indicates fair agreement. It suggests that Annotators 2 and 3 have a relatively more consistent understanding and application of the relevance criteria compared to the agreement between Annotator 1 and the others.

the scores show that the people marking documents as relevant or not don't always agree with each other. This could be because the rules they're following aren't clear enough, or they have their own opinions influencing their choices.